# <span style="color:gold">Algoritmi di Crittografia (2023/24)</span>
## Notebook 1

In [ ]:
from IPython.display import HTML
HTML('<style>{}</style>'.format(open('/home/mauro/.jupyter/custom/custom.css').read()))

Latex definitions
$\DeclareMathOperator*{\prob}{prob}$
$\DeclareMathOperator*{\mod}{mod}$
$\DeclareMathOperator*{\ln}{ln}$
$\def\zn{\mathbf{Z}_n}$
$\def\zp{\mathbf{Z}_p}$
$\def\mcd{\mathrm{MCD}}$

## <span style="color:gold">Presentazione del corso</span>

### <span style="color:cyan">Informazioni generali sul corso</span>
* 6 CFU, per <span style="font-style: italic; color:gold">42 ore</span> di lezione, divisi in due moduli:
    1. <span style="color:gold">Teoria e algoritmi</span> (4 CFU), prof. <span style="color:gold">Mauro Leoncini</span>
    2. <span style="color:gold">Crimine informatico</span> (2 CFU), prof. <span style="color:gold">Antonio Apruzzese</span>
    
<P />

* Il materiale didattico, in particolare i <span style="color:gold">notebook</span>, sarà disponibile su GDRIVE (il link di accesso verrà inviato ai partecipanti).
* Non è previsto un libro di testo. I <span style="color:gold">notebook</span> delle lezioni e gli appunti sul crimine informatico sono in linea di principio sufficienti per preparare l'esame.
* Tuttavia, agli interessati suggerisco almeno un paio di testi per lo studio e l'approfondimento personale, uno in lingua inglese e l'altro in lingua italiana.

<div style="text-align: center;">
    <div style="display: inline-block; text-align: left;">
        Jean-Philippe Aumasson<br />
        <span style="font-style: italic; color:gold">Serious Cryptography: A Practical Introduction to Modern Encryption</span><br />
        No Starch Press
<P />
        Anna Bernasconi, Paolo Ferragina, Fabrizio Luccio<br />
        <span style="font-style: italic; color:gold">Elementi di Crittografia</span><br />
        Pisa University Press
    </div>
</div>


* Esame solo <span style="font-style: italic; color:gold">orale</span>.
* Ricevimento studenti: <span style="color:gold">su richiesta</span> (a distanza o in presenza).

### <span style="color:cyan">Contenuti del corso (modulo Teoria e Algoritmi)</span>
* Cifrari classici ed elementi base della <span style="color:gold">crittografia simmetrica</span>
* Prerequisiti <span style="color:gold">matematico/algoritmici</span> per la crittografia asimmetrica
    * Casualità e <span style="color:gold">algoritmi probabilistici</span>
    * <span style="color:gold">Aritmetica modulare</span>
    * <span style="color:gold">Algoritmi di base</span> per la crittografia asimmetrica
* Scambio di chiavi con il <span style="color:gold">protocollo Diffie-Hellman</span>
* Protocolli di cifratura (<span style="color:gold">RSA</span>, <span style="color:gold">El Gamal</span>, <span style="color:gold">Rabin</span>)
* Firma digitale (<span style="color:gold">RSA</span>, <span style="color:gold">El Gamal</span>, <span style="color:gold">DSA</span>)
* <span style="color:gold">Curve ellittiche</span> e loro proprietà per applicazioni crittografiche
* <span style="color:gold">Firma digitale su curve ellittiche</span>
* Infrastrutture per la <span style="color:gold">certificazione</span> di chiavi pubbliche.

* Il corso sarà corredato:
    1. dalla presentazione (sviluppo e uso) di <span style="color:gold">algoritmi in linguaggio Python</span>
    2. dall'uso di <span style="color:gold">librerie crittografiche</span>

### <span style="color:cyan">Software utilizzato</span>

Utilizzeremo i seguenti strumenti software:
<ul>
    <li> il linguaggio <span style="font-style: italic; color:gold">Python 3.11</span>;</li>
    <li> l'ambiente <span style="font-style: italic; color:gold">jupyter</span> per le presentazioni a lezione (quello che stiamo usando ora);</li>
    <li> (forse, e occasionalmente) l'ambiente di sviluppo <span style="font-style: italic; color:gold">Spyder</span> </li>
</ul>

#### <span style="color:cyan">Installazione del software in ambiente virtuale sotto Linux</span>
* Le istruzioni qui riportate si riferiscono a sistemi basati su <span style="font-style: italic; color:gold">Debian GNU Linux</span> (Debian, Ubuntu, Mint, ...)
* Per altre distribuzioni quel che cambia è essenzialmente il <span style="color:gold">package manager</span>
```shell
sudo apt install python3-virtualenv
sudo apt install python3-virtualenvwrapper
mkdir ~/.virtualenvs
mkvirtualenv AC    # Il nome AC è arbitrario
workon AC
(AC) pip install spyder
(AC) pip install ipython
(AC) pip install jupyter-notebook
(AC) deactivate
```

<span style="color:cyan">Nota</span>: &Egrave; possibile che il comando <TT>mkvirtualenv</TT> non sia inizialmente riconosciuto. In tal caso, un workaround consiste nell'aggiungere le seguenti linee di codice al file <span style="color:gold"><TT></TT>.bashrc</span> nella propria home:
```shell
export VIRTUALENVWRAPPER_PYTHON=/usr/bin/python3
export ORKON_HOME=~/.virtualenvs
export ARCHFLAGS=’-arch x86_64’
source /usr/local/bin/virtualenvwrapper.sh
```

Attenzione inoltre alla possibilità che che il file <TT>virtualenvwrapper.sh</TT> non sia stato installato nella directory <TT>/usr/local/bin</TT>. In tal caso, lo si “localizzi” prima con il comando:
```shell
dpkg -L virtualenvwrapper | grep virtualenvwrapper.sh
```
e si modifichi poi opportunamente l'ultima riga da inserire in <TT>.bashrc</TT>.

#### <span style="color:cyan">Installazione sotto Windows o Mac</span>

* Per questi ambienti consigliamo di installare la piattaforma <span style="font-style: italic; color:gold">Anaconda</span> (che include <TT>IPython</TT>, <TT>jupyter-notebook</TT> e spyder).
* Anaconda è comunque disponibile <span style="color:gold">anche in ambiente Linux</span>.

#### <span style="color:cyan">Uso sotto GDrive</span>
* È possibile usare notebook dal proprio account Google, quindi <span style="color:gold">senza installare nulla</span> sul PC.
* È necessario accedere a Google drive e aprire i notebook (preventivamente caricati) usando l'applicazione <span style="font-style: italic; color:gold">COLAB</span> (COllaborative LABoratory)

### <span style="color:gold">Nozioni preliminari</span>

#### <span style="color:cyan">Scenario classico</span>
* La crittografia moderna include molteplici ambiti applicativi.
* Lo scenario "tipico" prevede due attori, tradizionalmente chiamati <span style="color:gold">Alice</span> e <span style="color:gold">Bob</span>, che vogliono scambiarsi informazioni su un <span style="color:gold">canale di comunicazione insicuro</span>, dove un terzo personaggio, <span style="color:gold">Eva</span>, è in grado di intercettare, leggere e modificare i messaggi.
* Assumendo che, in una data comunicazione, Alice sia il mittente e Bob il destinario, si possono individuare i seguenti <span style="color:gold">requisiti tipici</span> della comunicazione stessa.
    1. La <span style="color:gold">confidenzialità</span> del messaggio, ovvero che Eva non sia in grado di comprenderne il contenuto;
    2. l'<span style="color:gold">autenticazione</span> del mittente, in base alla quale Bob è certo che il messaggio sia stato effettivamente inviato da Alice;
    3. l'<span style="color:gold">integrità</span> del messaggio, in base alla quale Bob è certo che Eva non lo abbia manomesso;
    4. il <span style="color:gold">non ripudio</span>, per cui Alice non può negare, in un secondo momento, di aver inviato il messaggio a Bob.

#### <span style="color:cyan">Un po' di ulteriore terminologia</span>
* <span style="color:gold">Plaintext/testo in chiaro</span> è il messaggio che si vuole tenere segreto.
* <span style="color:gold">Encryption/cifratura</span> è il processo che rende il plaintext incomprensibile per Eva.
* Nella lingua inglese il testo cifrato è indicato con il termine <span style="color:gold">ciphertext</span>.
* Ogni processo di cifratura deve evidentemente essere <span style="color:gold">reversibile</span>.
* Il procedimento che dal testo cifrato restrituisce il plaintext originale è detto <span style="color:gold">decryption/decifrazione</span>.
* Cifratura e decifrazione sono i due algoritmi che compongono un <span style="color:gold">cifrario/cipher</span>
* Un cifrario fa uso di informazione segreta, nota come <span style="color:gold">chiave/key</span>

#### <span style="color:cyan">Il fondamentale principio di Kerckhoff</span>

<center>
    
    In uno schema di cifratura, la sicurezza deve risiedere
    solo nella segretezza della chiave
    (A. Kerckhoff, La Cryptographie Militaire, 1883)
</center>

* In particolare, la segretezza non deve risiedere nell'algoritmo.

#### <span style="color:cyan">Due "tipi" di crittografia</span>
* Letteralmente per "secoli" crittografia è stato solo un sinonimo di <span style="color:gold">crittografia simmetrica</span>.
* In un protocollo simmetrico Bob e Alice possiedono la stessa chiave ovvero hanno chiavi che possono essere <span style="color:gold">facilmente determinate</span> l'una a partire dall'altra.
* Solo in tempi recenti, sulla spinta della necessità di risolvere il problema cruciale dello <span style="color:gold">scambio di chiavi</span>, si è sviluppata la <span style="color:gold">crittografia asimmetrica</span>, in cui una delle due chiavi è posseduta solo da chi, fra Alice e Bob,
è il destinatario di una comunicazione.
* &Egrave; palese che, nel caso della crittografia asimmetrica, la determinazione della chiave privata a partire dalla chiave pubblica deve essere  <span style="color:gold">computazionalmente difficile</span>.
* Il lavoro fondamentale, che delinea le caratteristiche della crittografia asimmetrica, è del 1976:

<div style="text-align: center;">
    <div style="display: inline-block; text-align: left;">
        Whitfield Diffie, Martin Hellman<br />
        <span style="color:gold">New directions in cryptography</span><br />
        IEEE Trans. Inf. Theory (1976)
    </div>
</div>

* La matematica e l'algoritmica sottostanti i protocollo simmetrici e quelli simmetrici sono <span style="color:gold">sostanzialmente distinte</span>.

#### <span style="color:cyan">Sguardo d'insieme alle differenze fra i due tipi di crittografia</span>
<table style="margin: 0 auto; border: 1px solid black;border-collapse: collapse;">
    <tr>
        <th style="border: 1px solid black;"><div style="font-size: 16px; color:gold;">Crittografia simmetrica</div></th>
        <th style="border: 1px solid black;"><div style="font-size: 16px; color:gold;">Crittografia asimmetrica</div></th>
    </tr>
    <tr>
        <td style="border: 1px solid black;">
            <div style="font-size: 16px;">Una sola chiave per cifrare e decifrare
            </div>
        </td>
        <td style="border: 1px solid black;">
            <div style="font-size: 16px;">Una chiave pubblica per cifrare e una privata per decifrare
            </div>
        </td>
    </tr>
    <tr>
        <td style="border: 1px solid black;">
            <div style="font-size: 16px;">Il principale ostacolo è la necessità di un accordo preliminare sulla chiave
            </div>
        </td>
        <td style="border: 1px solid black;">
            <div style="font-size: 16px;">Il mittente deve essere certo dell'identit&agrave; del possessore della chiave pubblica
            </div>
        </td>
    </tr>
    <tr>
        <td style="border: 1px solid black;">
            <div style="font-size: 16px;">Ingestibile in applicazioni che coinvolgono più partecipanti (servono $O(n^2)$ chiavi)
            </div>
        </td>
        <td style="border: 1px solid black;">
            <div style="font-size: 16px;">Gestibile con più coppie di chiavi pubblica/privata (servono $n$ coppie)
            </div>
        </td>
    </tr>
    <tr>
        <td style="border: 1px solid black;">
            <div style="font-size: 16px;">Algoritmi veloci, anche con supporto hardware
            </div>
        </td>
        <td style="border: 1px solid black;">
            <div style="font-size: 16px;">Algoritmi lenti
            </div>
        </td>
    </tr>
</table>

#### <span style="color:cyan">Crittografia simmetrica e asimmetrica lavorano spesso insieme</span>
* Vantaggi e svantaggi sono speculari e dunque, molto spesso, protocolli applicativi che utilizzano la crittografia fanno uso di <span style="color:gold">entrambe le soluzioni</span>.
* Al riguardo, il seguente schema esemplifica le interazioni fra i due tipi di crittografia.
* Lo scenario è relativo alla <span style="color:gold">cifratura</span>.

<img src="cifratura_schema.jpg" width="500">

### <span style="color:gold">Crittografia simmetrica</span>
* Il focus di questo corso è sulla <span style="color:gold">crittografia asimmetrica</span>.
* In questo primo notebook forniamo tuttavia qualche nozione anche sulla <span style="color:gold">crittografia simmetrica</span>.
* I concetti base verranno introdotti utilizzando principalmente <span style="color:gold">cifrari classici</span>, del tutto inedeguati a resistere ad attacchi portati con moderni coputer.
* Vedremo tuttavia anche un cifrario moderno (anche se superato), che ci consentirà di cogliere le idee oggi alla base della crittografia simmetrica.

### <span style="color:cyan">Principi generali</span>
* Nella struttura di un cifrario simmetrico si possono individuare due distinte componenti.
* Una prima componente è l'algoritmo, o semplicemente l'insieme di operazioni che applicano una determinata <span style="color:gold">permutazione</span> ad una porzione del plaintext.
* Che cosa sia una porzione è evidentemente una caratteristica propria del cifrario.
* Essa puà essere una <span style="color:gold">singola lettera</span>, un gruppo di lettere o (come nei cifrari moderni) un <span style="color:gold">gruppo di bit</span>, detto comunemente <span style="color:gold">blocco</span>.
* Una <span style="color:gold">permutazione</span> di un insieme di $n$ oggetti è uno dei possibili modi in cui gli $n$ oggetti si possono disporre linearmente (uno dopo l'altro); cioè è uno dei possibili <span style="color:gold">ordinamenti</span> degli oggetti.
* In termini matematci, una permutazione (degli elementi) di un insieme $I$ è una funzione invertibile di $I$ in se stesso.
* Si noti che l'ovvio requisito di invertibilità delle cifratura, che si precisa formalmente con l'invertibilià delle permutazioni, in termini pratici implica che plaintext e ciphertext abbiano sempre la <span style="color:gold">stessa lunghezza</span>.
* La seconda componente fondamentale di un cifrario è l'algoritmo che, organizzando opportunamente le permutazioni dei singoli blocchi, effettua la <span style="color:gold">cifratura di un plaintext arbitrario</span>.
* Questa seconda componente è detta <span style="color:gold">mode of operation</span>.

<span style="font-style: italic; color:cyan">Nota</span>: i cifrari che seguono l'impostazione appena delineata sono detti, non a caso, (cifrari) <span style="color:gold">a blocchi</span>. Esistono però cifrari simmetrici differenti, che considerano il messaggio come <span style="color:gold">flusso continuo</span> di bit e che utilizzano la chiave per generare un  corrispondente flusso di bit (pseudo)casuali, da porre in XOR con i bit del messaggio. Tali cifrari, detti appunto <span style="font-style: italic; color:gold">stream cipher</span>, sono più adatti a, e vengono utilizzati per, cifrare (ad esempio) comunicazioni telefoniche.

#### <span style="color:cyan">Sulle permutazioni</span>
* &Egrave; noto che, al crescere di $n$, il numero di permutazioni diventa rapidamente enorme, in quanto cresce come il <span style="color:gold">fattoriale di $n$</span>.
$$
n! = 1\cdot2\cdot3\cdot\ldots\cdot(n-1)\cdot n
$$
* Ad esempio, le possibili permutazioni di 10 oggetti sono già più di 3 milioni. Numero che sale a circa 1300 miliardi nel caso di 15 oggetti.

In [ ]:
from math import factorial
print(factorial(5))
print(factorial(10))
print(factorial(26))

120
3628800
403291461126605635584000000


### <span style="color:cyan">Due cifrari classici: Cesare e Vigenère</span>
* Consideriamo messaggi scritti in un generico alfabeto $\Sigma$ e consideriamo un <span style="color:gold">ordinamento circolare dei caratteri</span> (arbitrario ma fissato una volta per tutte).
* Circolare vuol dire che il primo carattere è considerato il successore dell'ultimo.
* Nel <span style="color:gold">cifrario di Cesare</span> la cifratura avviene come illustrato di seguito.
    * La permutazione applicata ad ogni carattere consiste in uno <span style="color:gold">slittamento</span> (<span style="color:gold">shift</span>) di $k$ posizioni in avanti.
    * Questo vuol dire che ogni carattere $x$ del plaintext viene sostituito con il carattere che si trova <span style="color:gold">$k$ posizioni più avanti</span> rispetto ad $x$</span> nell'ordinamento circolare dell'alfabeto.
    * Ad esempio, considerando l'alfabeto inglese e ponendo $k=3$, la $a$ viene sostituita con la $d$, la $b$ con la $e$,..., la $x$ con la $a$, la $y$ con $b$ e la $z$ con la $c$.
    * Il mode of operation del cifrario è il più semplice possibile e consiste nell'applicare la <span style="color:gold">stessa permutazione</span> ad ogni carattere.
* &Egrave; evidente che chiave "segreta" di cifratura è proprio il valore <span style="color:gold">$k$</span> dello slittamento.
* La decifrazione deve rimettere le cose a posto e questo, nel cifrario di cesare, consiste nell'effettuare uno <span style="color:gold">slittamento</span> di $k$ posizioni all'indietro, ovvero sostituire ogni carattere $x$ del ciphertext con un carattere nell'alfabeto che <span style="color:gold">precede $x$ di $k$ posizioni</span> nell'ordinamento circolare.
* Potremmo anche dire che, se $k$ è la chiave di cifratura, la chiave di decifrazione è il valore opposto, <span style="color:gold">$-k$</span>.
* &Egrave; chiaro che per il cifrario di Cesare il <span style="color:gold">numero di chiavi</span>, che coincide con il numero di possibili slittamenti, è limitato dalla cardinalità dell'alfabeto.

#### <span style="color:cyan">Una semplice sfida</span>
* Consideriamo come alfabeto Σ = {a,b,...,z}  &cup; {.} &cup; {,}&cup; {'}
* Sapendo che il cifrario utilizzato è quello di Cesare, a quale testo in chiaro corrisponde il cyphertext <span style="color:gold">kdcxiitzz,ev,yhtjeved,bv,yhth,ew,vxithx</span>?

In [ ]:
alphabet = "abcdefghijklmnopqrstuvwxyz.,'"
def encrypt(plaintext,key):
    n = len(alphabet)
    cyphertext = ''.join([alphabet[(alphabet.find(c)+key)%n] for c in plaintext])
    return cyphertext
def decrypt(cyphertext,key):
    return encrypt(cyphertext,-key)

* Il procedimento di risoluzione di questa sfida è un caso di <span style="font-style: italic; color:gold">crittoanalisi</span>.
* In questo caso, la crittoanalisi è banale. Dato il numero limitato di chiavi, un <span style="color:gold">attacco a forza bruta</span> ha facile successo, solo con qualche necessaria sottolineatura...

In [ ]:
# Attacco a forza bruta al cifrario di Cesare
unknowntext = "kdcxiitzz,ev,yhtjeved,bv,yhth,ew,vxithx"
for k in range(len(alphabet)):
    print(f"{k}\t{decrypt(unknowntext,k)}")

0	kdcxiitzz,ev,yhtjeved,bv,yhth,ew,vxithx
1	jcbwhhsyy.du.xgsidudc.au.xgsg.dv.uwhsgw
2	ibavggrxxzctzwfrhctcbz'tzwfrfzcuztvgrfv
3	ha'uffqwwybsyveqgbsbay,syveqeybtysufqeu
4	g',teepvvxarxudpfara'x.rxudpdxasxrtepdt
5	f,.sddouuw'qwtcoe'q',wzqwtcocw'rwqsdocs
6	e.zrccnttv,pvsbnd,p,.vypvsbnbv,qvprcnbr
7	dzyqbbmssu.ouramc.o.zuxouramau.puoqbmaq
8	cyxpaalrrtzntq'lbznzytwntq'l'tzotnpal'p
9	bxwo''kqqsymsp,kaymyxsvmsp,k,synsmo'k,o
10	awvn,,jpprxlro.j'xlxwrulro.j.rxmrln,j.n
11	'vum..iooqwkqnzi,wkwvqtkqnzizqwlqkm.izm
12	,utlzzhnnpvjpmyh.vjvupsjpmyhypvkpjlzhyl
13	.tskyygmmouiolxgzuiutoriolxgxoujoikygxk
14	zsrjxxfllnthnkwfythtsnqhnkwfwntinhjxfwj
15	yrqiwwekkmsgmjvexsgsrmpgmjvevmshmgiwevi
16	xqphvvdjjlrfliudwrfrqlofliudulrglfhvduh
17	wpoguuciikqekhtcvqeqpknekhtctkqfkeguctg
18	vonfttbhhjpdjgsbupdpojmdjgsbsjpejdftbsf
19	unmessaggiocifratoconilcifrariodicesare
20	tmldrr'ffhnbheq'snbnmhkbheq'qhnchbdr'qd
21	slkcqq,eegmagdp,rmamlgjagdp,pgmbgacq,pc
22	rkjbpp.ddfl'fco.ql'lkfi'fco.oflaf'bp.ob
23	qjiaoozccek,ebnzpk

#### <span style="color:cyan">Cifrari mono-alfabetici</span>
* Il <span style="color:gold">cifrario di Cesare</span> è forse il più semplice caso di <span style="color:gold">cifrario mono-alfabetico</span>.
* &Egrave; tale un cifrario che utilizza una <span style="color:gold">sostituzione fissa</span> per ogni carattere.
* In altri termini, fissato l'alfabeto di riferimento $\Sigma$, un cifrario mono-alfabetico è definito da una <span style="color:gold">singola permutazione</span> ripetuta per tutti i caratteri.

#### <span style="color:cyan">Permutazioni ideali e relazioni con la chiave</span>

* Le permutazioni utilizzate dal cifrario di Cesare sono semplici slittamenti circolari, detti più propriamente <span style="color:gold">rotazioni</span>.
* Per conoscere una tale permutazione è sufficiente un singolo numero intero, compreso fra 0 e la cardinalità dell'alfabeto, che rappresenta quindi la <span style="color:gold">chiave del cifrario</span>.
* La proprietà che la chiave determina la permutazione deve valere <span style="color:gold">sempre</span>, anche se il passaggio dalla chiave alla permutazione non è di norma così semplice come nel cifrario di Cesare.
* La chiave infatti è l'<span style="color:gold">unica informazione segreta</span> e dunque la permutazione deve dipendere necessariamente dalla chiave.
* Non solo, chiavi diverse devono sempre fornire permutazioni diverse. In caso contrario per un attaccante ci sarebbero <span style="color:gold">meno chiavi distinte da "provare"</span> e questo potrebbe agevolare sensibilmente gli attacchi.
* Infine, la permutazione dovrebbe il più possibile apparire <span style="color:gold">casuale</span>. Questo garantisce che la conoscenza di come viene mappata una lettera (o un gruppo di bit) non fornisce alcuna conoscenza sulla mappatura delle altre.
* Nel caso del cifrario di Cesare, la permutazione è tutto fuorché casuale. Se sappiamo che la lettera $a$ viene mappata nella lettera $d$, sappiamo immediatamente anche che $b$ viene mappata in $e$, $c$ in $f$ e così via.

#### <span style="color:cyan">Un generico cifrario mono-alfabetico</span>
* Allo scopo di procurarci permutazioni arbitrarie, utilizzeremo la funzione <span style="color:gold"><TT>permutation</TT></span> del modulo <span style="color:gold"><TT>numpy.random</TT></span>.

In [ ]:
from numpy.random import permutation
p=permutation(5)
print([i for i in range(5)])
print([i for i in p])

[0, 1, 2, 3, 4]
[4, 0, 1, 3, 2]


* In questo esempio usiamo il seguente alfabeto, composto di 30 simboli

In [ ]:
alphabet2 = r"abcdefghijklmnopqrstuvwxyz .,'"
factorial(30)

265252859812191058636308480000000

* Senza starle a contare, il numero di cifre decimali $30!$ è determinabile nel modo seguente

In [ ]:
from math import log10
X = factorial(30)
int(log10(X))+1

33

In [ ]:
# Funzione di cifratura
def mono_alphabetic_encrypt(plaintext,alphabet,P=None):
    '''Uses P or otherwise picks a random key/permutation and
       encypts plaintext over alphabet. Returns cyphertext and key'''
    from numpy.random import permutation
    # Possiamo ignorare le accentate perché questo non inficia l'intellegibilità
    # ad un madre-lingua italiano
    D = {'à':'a','è':'e','é':'e','ì':'i','ò':'o','ù':'u'}
    if P is None:
        P=permutation(len(alphabet))
    ciphertext = [alphabet[P[alphabet.find(D.get(c,c))]] for c in plaintext]
    return ''.join(ciphertext),P

In [ ]:
c,k = mono_alphabetic_encrypt("Sopra la panca, la capra a da campà", alphabet2)
print(c)

rnkmptdptkpzepwtdptepkmptptaptepbkp


* La chiave ovviamente <span style="color:gold">non può viaggiare in chiaro</span>; deve in qualche modo essere preliminarmente concordata.
* Come esercizio, si scriva il semplice codice per la decifrazione.
* Al riguardo, il seguente codice calcola la <span style="color:gold">permutazione inversa</span> di una permutazione data come array <TT>numpy</TT>.

In [ ]:
from numpy import where

In [ ]:
# Una generica permutazione
P = permutation(5); P

In [ ]:
# La corrispondente permutazione inversa
Pinv = [where(P==i)[0][0] for i in range(len(P))]; Pinv

In [ ]:
# Verifica
for i in range(len(P)):
    print(Pinv[P[i]],end='\t')

#### <span style="color:cyan">Una seconda challenge: decifrare il testo cifrato che stiamo per creare</span>

In [ ]:
# hiddentext.txt contiene un testo in italiano da decifrare
P = permutation(len(alphabet2))
with open('hiddentext.txt','r') as f:
    ciphertext,key = \
    mono_alphabetic_encrypt(''.join([l.strip().lower() for l in f.readlines()]),alphabet2,P)

In [ ]:
print(ciphertext)

bfqroorvlja zbbdapjttrbaxjvafqasrvvjalfapjfwafqasrvvjalfax faszaljkrasfazvvjlrbzwazqqzacfvrwajovuvjalrfacfqfax ra.ninabj,qfvoa zafvxjyfvxfztjazatrssrbralzqqzapbfyzapzofvzalrqapbfyjakjquyrwafqaxjyftztjasfaratbjkztjalfacbjvtrwaxjyravrqqracfz'raravrfaqz'fbfvtfwatbrakfrnayzvtrvrbraqzasxrqtzaczttzavrqatbzlubbravjyfalfaqujo faraprbsjvzoofwabftjbvzbrazqavjyrajbfofvzqrajaprbxjbbrbrauvzatrbmzakfzasxroqfrvljauvavjyralrqatuttjavujkjnazqxuvfarsrypfnvrqaxzsjalfazq'usasfqrvtrarapzbsjasu'ftjax fzbjax razapbfjbfavrssuvzalrqqratbrastbzlrarbzalfaprbasraeurqqzaofustznazqayjyrvtjalfasxroqfrbrafqaxjovjyraftzqfzvjwax rarbzapzbsjazlrouztjaprbauvayzoja'fmmzbbjayzazvx rasjqrvvraraxzpzxralfatrvrbrafvasjoormfjvrafasujfavryfxfwavjvasfaszprkzaeurqqjax ra.ninabj,qfvoazkbr''rapjfalfx fzbztjagqjafyyzofvzkjaxjyrauvayzoja'rvrkjqjwasrypbrafvayjkfyrvtjwax rayjbyjbzaxjvtfvuzyrvtratbzasrarasrgaluy'qrljbrwafvafvoqrsrwarafqavjyrazbxzfxjalfa'uy'qr'rrwafqaxzqz'bjvrnazqtbjax ragsfqrvtrgnarppubrwaqzastjbfzalfyjstbrbzax rapbjpbfj

* Sembra una challenge impossibile, visto che la permutazione è stata scelta a caso fra <span style="color:gold">30! permutazioni</span> possibili.

#### <span style="color:cyan">Critto-analisi</span>
* Per un generico cifrario mono-alfabetico il numero di possibili chiavi è dunque pari al numero di permutazioni dell'insieme dei caratteri dell'alfabeto.
* Per l'alfabeto inglese, senza spazi ne' caratteri di punteggiatura, abbiamo abbiamo quindi <span style="color:gold">$26!\approx 4\cdot 10^{26}\approx 1.3\cdot 2^{88}$</span> permutazioni</span>. Tale numero è paragonabile al numero di atomi nel corpo umano.
* Un attacco a forza bruta è pertanto <span style="color:gold">fuori discussione</span>.
* Per i cifrari mono-alfabetici è però efficace una critto-analisi <span style="color:gold">basata sull'analisi delle frequenze</span>.

#### <span style="color:cyan">Analisi della frequenze</span>
* L'attacco è basato sull'ipotesi (non sempre verificata) che la frequenza con cui compaiono i singoli caratteri nel plaintext sia conforme alla più generale <span style="color:gold">frequenza nella lingua</span> con cui il testo stesso è scritto.
* &Egrave; evidente che, in generale, l'ipotesi sarà tanto più verificata quanto più il plaintext è lungo.
* Per portare questo attacco, conoscendo (o ipotizzando quale sia) la lingua con cui è scritto il plaintext, bisogna dunque disporre di un <span style="color:gold">stima delle frequenze dei caratteri</span> in quella lingua.
* Le celle di codice che seguono calcolano una tale stima per l'italiano, utilizzando a tale scopo un solo testo "classico" (<span style="color:gold">I promessi sposi</span>).

In [ ]:
def countfreq(text, alphabet):
    '''Given some text, computes the frequencies of the characters included
       in the alphabet. Returns a dictionary. It is assumed that the
       alphabet includes lowercase letters only.
       text is a string but a first attempt is made to check whether
       filename is the complete name of a file storing the actual text.
    '''
    try:
        f = open(text,'r')
        T = ''.join([line.strip().lower() for line in f.readlines()])
        f.close()
    except OSError or FileNotFoundError:
        # First parameter is the actual text
        T = text.lower()
    D = {c:0 for c in alphabet}
    for c in T:
        c = {'à':'a','è':'e','é':'e','ì':'i','ò':'o','ù':'u'}.get(c,c)
        try:
            D[c]+=1
        except KeyError:
            # Discard not-alphabetic characters
            pass
    return D

In [ ]:
countfreq("tanto va la gatta al lardo che ci lascia lo zampino",alphabet2)

{'a': 10,
 'b': 0,
 'c': 3,
 'd': 1,
 'e': 1,
 'f': 0,
 'g': 1,
 'h': 1,
 'i': 3,
 'j': 0,
 'k': 0,
 'l': 5,
 'm': 1,
 'n': 2,
 'o': 4,
 'p': 1,
 'q': 0,
 'r': 1,
 's': 1,
 't': 4,
 'u': 0,
 'v': 1,
 'w': 0,
 'x': 0,
 'y': 0,
 'z': 1,
 ' ': 10,
 '.': 0,
 ',': 0,
 "'": 0}

In [ ]:
Fdict=countfreq('promessi_sposi.txt',alphabet2)
freq = sorted([(k,v) for k,v in Fdict.items()] ,key=lambda p: p[1],reverse=True)

In [ ]:
freq

[(' ', 196150),
 ('e', 123641),
 ('a', 118130),
 ('o', 98961),
 ('i', 97929),
 ('n', 74801),
 ('r', 67744),
 ('t', 62555),
 ('l', 57139),
 ('s', 56060),
 ('c', 48112),
 ('d', 38214),
 ('u', 36592),
 ('p', 30455),
 (',', 26319),
 ('m', 24275),
 ('v', 23585),
 ('g', 17585),
 ('h', 13834),
 ('f', 10804),
 ('b', 10009),
 ("'", 9955),
 ('.', 9553),
 ('q', 8007),
 ('z', 7776),
 ('x', 75),
 ('j', 23),
 ('y', 8),
 ('w', 3),
 ('k', 0)]

* Come secondo passo calcoliamo le <span style="color:gold">frequenze dei caratteri ciphertext</span> e ordiniamole in senso decrescente.

In [ ]:
cFdict=countfreq(ciphertext,alphabet2)
cfreq = sorted([(k,v) for k,v in cFdict.items()] ,key=lambda p: p[1],reverse=True)

In [ ]:
cfreq

[('a', 295),
 ('r', 200),
 ('z', 151),
 ('f', 148),
 ('j', 138),
 ('v', 116),
 ('b', 104),
 ('q', 99),
 ('t', 87),
 ('s', 75),
 ('x', 54),
 ('l', 49),
 ('y', 43),
 ('u', 40),
 ('o', 38),
 ('p', 38),
 ('k', 27),
 ('w', 26),
 (' ', 20),
 ('n', 18),
 ("'", 14),
 ('c', 9),
 ('m', 9),
 ('e', 6),
 ('g', 6),
 (',', 3),
 ('d', 2),
 ('i', 2),
 ('.', 2),
 ('h', 0)]

* Come terzo passo proviamo la <span style="color:gold">corrispondenza (permutazione) secca</span>.
* Dopodiché, se si inizia a "vedere qualcosa", si eseguono modifiche progressive alla permutazione iniziale.

In [ ]:
# Permutazione iniziale
invkey = {p[0]:q[0] for p,q in zip(cfreq,freq)}

In [ ]:
# Applicazione della permutazione attuale
ptext = ''.join([invkey[c] for c in ciphertext])
print(ptext)

rileggendo harry potter con il senno di poi, il senno di chi sa dove si annodera, alla fine, ognuno dei fili che w'j' roxling ha incominciato a tessere dalla prima pagina del primo volume, il comitato si e trovato di fronte, come nelle fiabe e nei labirinti, tre vie' mantenere la scelta fatta nel tradurre nomi di luoghi e personaggi, ritornare al nome originale o percorrere una ter.a via scegliendo un nome del tutto nuovo' alcuni esempi'nel caso di albus silente e parso subito chiaro che a priori nessuna delle tre strade era di per se quella giusta' al momento di scegliere il cognome italiano, che era parso adeguato per un mago bi..arro ma anche solenne e capace di tenere in sogge.ione i suoi nemici, non si sapeva quello che w'j' roxling avrebbe poi dichiarato zlo immaginavo come un mago benevolo, sempre in movimento, che mormora continuamente tra se e sez dumbledore, in inglese, e il nome arcaico di bumblebee, il calabrone' altro che zsilentez' eppure, la storia dimostrera che proprio

In [ ]:
# Funzione che inverte la corrispondenza di v1 e v2
def swap(v1,v2,d):
    k1 = list(d.keys())[list(d.values()).index(v1)]
    k2 = list(d.keys())[list(d.values()).index(v2)]
    d[k1] = v2
    d[k2] = v1

* Dall'<span style="color:gold">analisi visuale</span> si effettuano poi i seguenti cambiamenti alla permutazione iniziale
* Si noti che, se si capisce che due lettere sono semplicemente invertite, l'applicazione di $\mathtt{swap}$ sistema entrambe le corrispondenze.
* Se invece una lettera $x$ del testo cifrato codifica $y$ del testo in chiaro ma $y$ non codifica $x$, allora $\mathtt{swap}(x,y,\mathtt{invkey})$ sistema solo la corrispondenza di $y$.
* In questo secondo caso, dopo lo swap la lettera che era codificata da $y$ ora è codificata sa $x$.
* I passaggi sono (relativamente) tanti, ma via via <span style="color:gold">sempre più ovvi</span>.

In [ ]:
swap("m","u",invkey)   # i e o sembrano invertite

In [ ]:
invkey

{'a': ' ',
 'r': 'e',
 'z': 'a',
 'f': 'i',
 'j': 'o',
 'v': 'n',
 'b': 'r',
 'q': 't',
 't': 'l',
 's': 's',
 'x': 'c',
 'l': 'd',
 'y': 'u',
 'u': 'p',
 'o': ',',
 'p': 'm',
 'k': 'v',
 'w': 'g',
 ' ': 'h',
 'n': 'f',
 "'": 'b',
 'c': "'",
 'm': '.',
 'e': 'q',
 'g': 'z',
 ',': 'x',
 'd': 'j',
 'i': 'y',
 '.': 'w',
 'h': 'k'}

#### <span style="color:cyan">Vigenère: un cfrario poli-alfabetico</span>

* Il cifrario di <span style="font-style: italic; color:gold">Vigenère</span> è un esempio di cifrario poli-alfabetico.
* Questo vuol dire che uno stesso carattere non è sempre soggetto alla stessa trasformazione.
* Le chiavi sono qui <span style="color:gold">sequenze di caratteri dell'alfabeto</span>, ciascuno dei quali deve essere interpretato come il numero corrispondente alla sua <span style="color:gold">posizione nell'alfabeto</span> (partendo da 0).
* Limitandoci ancora una volta all'alfabero inglese, possiamo dunque affermare che, ad esempio, la chiave <span style="color:gold"><TT>crypto</TT></span>, vada interpretata come la sequenza di numeri <span style="color:gold"><TT>2,17,24,15,19,14</TT></span> (dove abbiamo usato le virgole per evitare possibili ambiguità).
* Una volta operata questa corrispondenza, la chiave viene usata nel modo seguente:
    * Il primo carattere del testo viene fatto slittare di 2 posizioni
    * Il secondo carattere del testo viene fatto slittare di 17 posizioni
    * ...
    * Il sesto carattere del testo viene fatto slittare di 14 posizioni
    * Il settimo carattere del viene fatto slittare nuovamente di 2 posizioni
    * ...

* Ad esempio, se il plaintext fosse <span style="color:gold"><TT>algorithms</TT></span> (e la chiave sempre <span style="color:gold"><TT></TT>crypto</span>), il ciphertext viene determinato nel modo seguente:

\begin{array}{rcccccccccc}
    \textrm{key}:&\mathtt{c}&\mathtt{r}&\mathtt{y}&\mathtt{p}&\mathtt{t}&\mathtt{o}&\mathtt{c}&\mathtt{r}&\mathtt{y}&\mathtt{p}\\
    \textrm{plaintext}:&\mathtt{a}&\mathtt{l}&\mathtt{g}&\mathtt{o}&\mathtt{r}&\mathtt{i}&\mathtt{t}&\mathtt{h}&\mathtt{m}&\mathtt{s}\\
    \textrm{ciphertext}:&\mathtt{c}&\mathtt{c}&\mathtt{e}&\mathtt{d}&\mathtt{k}&\mathtt{w}&\mathtt{v}&\mathtt{y}&\mathtt{k}&\mathtt{h}
\end{array}

* La decifrazione procede in modo analogo, usando la stessa chiave ma con <span style="color:gold">slittamenti all'indietro</span>.

* Il cifrario di Vigenère può essere analizzato in termini generali (e non come "sequenza" di cifrari di Cesare).
* Nel caso di Vigenère le porzioni di testo sottoposte a permutazione sono <span style="color:gold">sequenze di $k$ lettere</span>, dove $k$ questa volta è la lunghezza della chiave.
* La chiave permette dunque di calcolare la cifratura univoca di qualsiasi <span style="color:gold">blocco</span> composto da $k$ lettere.
* La permutazione è dunque rappresentabile per esteso mediante una <span style="color:gold">tabella di $26^k$ righe</span>, una per ogni possibile blocco.
* L'esempio seguente mostra le righe iniziali e finali della tabella corrispondente alla chive <TT>crypto</TT>, la cui lunghezza (numero di righe) è $26^6=308915776$.


<table style="margin: 0 auto; border: 1px solid black;border-collapse: collapse;">
       <tr>
               <th style="border: 1px solid black;text-align:center"><div style="font-size: 16px; color:gold;">plaintext</div></th>
           <th style="border: 1px solid black;text-align:center"><div style="font-size: 16px; color:gold;">ciphertext</div></th>
    </tr>
       <tr>
               <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>aaaaaa</TT></div></td>
           <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>crypto</TT></div></td>
    </tr>
    <tr>
               <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>aaaaab</TT></div></td>
           <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>cryptp</TT></div></td>
    </tr>
    <tr>
               <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>aaaaac</TT></div></td>
           <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>cryptq</TT></div></td>
    </tr>
    <tr>
               <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>...</TT></div></td>
           <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>...</TT></div></td>
    </tr>
    <tr>
               <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>yzzzzz</TT></div></td>
           <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>aqxosn</TT></div></td>
    </tr>
        <tr>
               <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>zzzzzz</TT></div></td>
           <td style="border: 1px solid black;text-align:center"><div style="font-size: 16px;"><TT>bqxosn</TT></div></td>
    </tr>
    </table>

#### <span style="color:cyan">Attacchi al cifrario di Vigenère</span>
* Senza l'ausilio dei computer, il cifrario era pressoché inattaccabile, in caso di chiave essenzialmente <span style="color:gold">casuale</span>.
* Rimane tutt'oggi difficilmente attacabile in caso di chiave casuale, di lunghezza confrontabile con il testo e non riutilizzata più volte.
* In quest'ultimo caso, il cifrario "degenera" infatti in una sorta di <span style="color:gold">one-time pad testuale</span> e il one-time pad, detto anche <span style="font-style: italic; color:gold">cifrario di Vernam</span>, è l'unico matematicamente inviolabile (come vedremo subito).
* Se invece la chiave è relativamente corta rispetto al testo è possibile portare un attacco, la cui "prima fase" consiste proprio nell'individuazione della <span style="color:gold">lunghezza della chiave</span>.

* Si consideri, come esempio guida, il testo cifrato <span style="color:gold">$C=\mathtt{fxbxzktwrdvfcsfxbxqwplbn}$</span>.
* Notiamo che la sequenza di 4 caratteri <span style="color:gold">$\mathtt{fxbx}$</span> si ripete ad una distanza di 14 posizioni.
* La ripetizione può certamente essere l'identica "risultante" di differenti caratteri in testo e chiave, ma è pure possibile che essa sia dovuta ad una <span style="color:gold">stessa sequenza nel testo in chiaro</span>.
* Se questo è il caso, allora devono però essere uguali <span style="color:gold">anche i caratteri della chiave</span> in quelle posizioni.
* Ma allora, continuando con le deduzioni, o la chiave è lunga <span style="color:gold">almeno 18 caratteri</span>, con al suo interno delle ripetizioni, oppure la chiave ha una lunghezza che è un <span style="color:gold">divisore di 14</span>, il che, in questo caso, vuol dire lunghezza 2 o 7 o 14.
* Partendo da queste osservazioni, l'attaccante può tentare <span style="color:gold">attacchi esaustivi con le lunghezze ammissibili più corte</span> (sicuramente 2 ma anche 7, se l'alfabeto non è troppo grande) o attacchi <span style="color:gold">basati su dizionario</span>, nel caso delle lunghezze superiori.

In [ ]:
int(log10(256**7))+1

17

#### <span style="color:cyan">Ipotesi "di lavoro da parte di Eva": la chiave è lunga 7</span>
* Eva può tentare (almeno) 3 strade:
    * Attacco <span style="color:gold">brute force</span>; pesante ma non impossibile perché le chiavi sono "soltanto" $26^7=8031810176$, cioè approssimativamente 8 miliardi.
    * <span style="color:gold">Analisi delle frequenze</span>: se il testo è abbastanza lungo Eva può stimare le frequenze considerando che due identici caratteri del testo cifrato corrispondono con certezza allo stesso carattere del plaintext se e solo se la loro posizione differisce di un multiplo di 7.
    * Attacco <span style="color:gold">basato sul dizionario</span>: ipotizzando che la chiave non sia una sequenza casuale, Eva può provare tutte le parole di 7 lettere in un dizionario inglese.
* Se l'attacco basato sul dizionario è chiaramente "parato" mediante l'utilizzo di una chiave casuale, l'analisi delle frequenze richiede una <span style="color:gold">buona dose di fortuna</span>.
* Ad esempio, anche il carattere più frequente potrebbe non essere individuato se esso viene trasformato in molti caratteri diversi a causa di allineamenti "sfortunati" (per Eva!)

#### <span style="color:cyan">Attacco con l'uso del dizionario</span>
* Supponiamo di sapere che la lingua del messaggio è l'inglese e che l'alfabeto sia
$$
\Sigma = \{\,\,,\mathtt{a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z}\}
$$
* Usiamo l'elenco di parole presenti nel file <span style="color:gold"><TT>cracklib-small</TT></span> del pacchetto Debian <span style="color:gold"><TT>cracklib-runtime</TT></span>.
* Nel file ci sono circa <span style="color:gold">8K parole di 7 caratteri</span> e un attaccante cercherà sicuramente una chiave nel dizionario.

* Vediamo (pur nella semplicità dell'esempio) come eseguire l'attacco
* Come prima cosa recuperiamo le parole di 7 lettere dal dizionario

In [ ]:
# /usr/share/dict/cracklib-small è un dizionario sul "mio computer". Può essere
# necessario modificare il path
with open('/usr/share/dict/cracklib-small', 'r') as f:
    W = {b.lower() for w in f if (b := w.strip().split("'")[0]) and len(b)==7}
print(len(W))

8579


In [ ]:
list(W)[1000:1020]

['tonight',
 'drywall',
 'minuend',
 'outgrow',
 'pitbull',
 'polices',
 'fiddles',
 'dastard',
 'bastion',
 'rivulet',
 'tabloid',
 'touched',
 'bounced',
 'drumlin',
 'gryphon',
 'bearing',
 'hermann',
 'hosting',
 'sudsing',
 'dullest']

* Dopodichè definiamo solo l'operazione fondamentale per eseguire la decifrazione, ovvero lo slittamento all'indietro

In [ ]:
alphabet3 = " abcdefghijklmnopqrstuvwxyz"
def leftrotate(x,y,alphabet):
    return alphabet[(alphabet.find(x)-alphabet.find(y))%len(alphabet)]

In [ ]:
leftrotate('c','h',alphabet3) # ruota 'c' di find('h') posizioni verso sx

'v'

* Per evitare di dover controllare circa 8K possibili soluzioni, utilizziamo un modulo Python in grado di rilevare il linguaggio di una sequenza di testo

In [ ]:
from langdetect import detect_langs,detect

* Possiamo ora tentare l'attacco

In [ ]:
T = "fxbxzktwrdvfcsfxbxqwplbn"
N = len(T)
candidates = []
Q,R = N//7,N%7
for w in W:
    key = w*Q+w[:R]
    plaintext = ''.join([leftrotate(T[i],key[i],alphabet3) for i in range(N)])
    lang = detect_langs(plaintext)
    if len(lang)==1 and lang[0].lang=='en' and lang[0].prob>0.9999:
        candidates.append((plaintext,w))

In [ ]:
len(candidates)  # Una buona riduzione da poco più di 8K

123

* L'analisi finale dei <span style="color:gold">plaintext candidati</span> può essere fatta da un essere umano.

In [ ]:
for c in sorted(candidates,key=lambda p:p[1]):
    print(c)

('evqwffpvpsumyoevqwxrlk b', 'ablated')
('evqwffavpsumy evqwxrxk b', 'ablates')
('cwotqfptqqrxyocwothrlia ', 'candied')
('cwotnfatqqruy cwoterxia ', 'candles')
('cwhefbqtqjcmupcwhexnmiat', 'caustic')
('cwgsmjftqiqtbecwgsdvbias', 'caveman')
('cpafffptjcdmyocpafxrlium', 'charted')
('cphflfptjjdsyocphfcrliut', 'churned')
('computational complexity', 'ciphers')
('clxwhfptfzuoyoclxwzrliqi', 'cleared')
('clxfofptfzdvyoclxffrliqi', 'clerked')
('clthjfptfvfqyocltharliqe', 'clipped')
('clthjfbtfvfqyacltharyiqe', 'clipper')
('ciphytotcrfelnciphpekina', 'compare')
('ciphnfwtcrfuyvciphersina', 'complex')
('ciphksotcrfrknciphbdkina', 'compose')
('ciphks tcrfrkzciphbdwina', 'compost')
('ciphfwftcrfmoeciphxhbina', 'compton')
('cipherotcrfljnciphwckina', 'compute')
('ciouypotcqsehncioupakin ', 'concave')
('ciouut tcqsalzcioulewin ', 'concert')
('ciouktptcqsrlocioubelin ', 'concord')
('cihurfptcjsyyocihuirlint', 'couched')
('cihlffbtcjjmyacihlxryint', 'coulter')
('cihhnfptcjfuyocihherlint', 'coupled')


### <span style="color:cyan">One-time pad</span>
* One time pad è "teoricamente" il cifrario <span style="color:gold">più sicuro</span>, anche se è affetto da molteplici problemi di natura pratica che lo rendono di fatto inutilizzabile.
* Il messaggio (plaintext) $P$ da cifrare è una <span style="color:gold">stringa di bit</span> (ed è naturalmente sempre possibile considerarlo tale).
* La chiave è una sequenza di bit della <span style="color:gold">stessa lunghezza del testo in chiaro</span>: l'$i$-esimo bit $k_i$ della chiave viene messo in xor con il corrispondente bit $p_i$ del plaintext per produrre l'$i$-esimo bit $c_i$ del ciphertext
$$
    c_i = p_i\oplus k_i,\qquad i=0,1,\ldots,|P|-1
$$
* Un semplice esempio
\begin{array}{rcccccccccc}
\textrm{key}:&0&0&0&1&0&1&1&0&1&0\\
        \textrm{p}:  &1&0&1&1&0&0&1&0&0&1\\
        \textrm{c}:  &1&0&1&0&0&1&0&0&1&1
\end{array}


* Per le proprietà del'or esclusivo, la decifrazione procede allo <span style="color:gold">stesso modo</span>, eseguendo cioè l'operazione con chiave e ciphertext:
\begin{array}{rcccccccccc}
\textrm{key}:&0&0&0&1&0&1&1&0&1&0\\
\textrm{c}:  &1&0&1&0&0&1&0&0&1&1\\
\textrm{p}:  &1&0&1&1&0&0&1&0&0&1
\end{array}

#### <span style="color:cyan">Scelta della chiave</span>
* Ogni bit della chiave deve avere valore 0 oppure 1 con <span style="color:gold">uguale probabilità</span> e in modo <span style="color:gold">indipendente</span> dal valore dei bit precedenti
$$
\prob[k_i=0]=\prob[k_i=1]=\frac{1}{2}\quad \prob[k_i=v|k_{i-1}=w]=\prob[k_i=v],\,\,v,w\in\{0,1\}
$$
* In questo modo, è immediato osservare che per i bit del messaggio cifrato <span style="color:gold">valgono le stesse proprietà</span>, indipendentemente da come sono scelti i bit del messaggio.
* Per la generica posizione (tralasciamo quindi gli indici per semplicità) vale infatti
$$
\begin{eqnarray*}
\prob[c = 1]&=&\prob[p\oplus k = 1]\\
&=& \prob[p=1 \wedge k=0]+\prob[p=0 \wedge k=1]\\
&=& \prob[p=1]\cdot\prob[k=0]+\prob[p=0]\cdot\prob[k=1]\\
&=& \alpha\cdot 0.5+(1-\alpha)\cdot 0.5\\
&=&0.5
\end{eqnarray*}
$$
dove abbiamo indicato con $\alpha$ la probabilità che il bit del messaggio sia 1. Lo stessa conclusione vale ovviamente per la probabilità che il bit del testo cifrato sia 0.
* Si noti, in particolare, che l'uguaglianza precedente vale anche se $\alpha=1$ o $\alpha=0$, cioè se il bit (e più in generale il messaggio) è <span style="color:gold">deterministico</span>.

#### <span style="color:cyan">Sicurezza del one-time pad</span>
* Se la chiave è scelta come indicato sopra ed $n$ è la lunghezza del plaintext, ogni messaggio cifrato ha la <span style="color:gold">stessa probabilità $2^{-n}$ di essere trasmesso</span>.
* Questa è ovviamente il massimo di garanzia perché, rovesciando la prospettiva, per l'attaccante il messaggio "visto" può corrispondere ad <span style="color:gold">uno qualsiasi dei $2^n$ possibili messaggi in chiaro</span>, tutti con identica probabilità $2^{-n}$.
* Tuttavia, se la chiave viene utilizzata più volte, l'attaccante può <span style="color:gold">ottenere conoscenza sui messaggi</span> (anche se non direttamente i messaggi in chiaro).


* Supponiamo, ad esempio, che due messaggi, $C_1$ e $C_2$, siano stati <span style="color:gold">cifrati usando la stessa chiave $K$</span>. Indicando con $O$ la stringa di tutti $0$ della stessa lunghezza della chiave, abbiamo
$$
\begin{eqnarray*}
C_1\oplus C_2&=&(P_1\oplus K)\oplus(P_2\oplus K)\\
&=&(P_1\oplus P_2)\oplus(K\oplus K)\\
&=&(P_1\oplus P_2)\oplus O\\
&=&P_1\oplus P_2
\end{eqnarray*}
$$

* In questo modo l'attaccante viene a conoscenza dello <span style="color:gold">xor dei due messaggi</span> e dunque, conoscendone uno dei due (o parte di uno dei due) verrebbe a conoscere anche (parte de) l'altro.

### <span style="color:cyan">Obiettivi generali di sicurezza</span>

* La precedente osservazione riguardo one-time pad ci porta a parlare degli <span style="color:gold">obiettivi generali di sicurezza</span>, almeno nel contesto della cifratura.
* Il primo obiettivo è, ovviamente, che Eva non sia in grado di comprendere i messaggi. Tuttavia questa è una formulazione <span style="color:gold">troppo generica</span>.
* Esistono <span style="color:gold">più obiettivi di attacco</span> da parte di Eva, di cui "capire il messaggio" non è neppure il più estremo.
* Ad esempio, <span style="color:gold">determinare la chiave</span> è ancora più ambizioso perché consente di mettere in chiaro uno o più messaggi.
* Un altro obiettivo consiste nel far passare per autentico un messaggio che invece Eva ha <span style="color:gold">alterato</span> o prodotto nella sua interezza.

#### <span style="color:cyan">Indistinguibilità e non-malleabilità</span>
* L'<span style="color:gold">indistinguibilità</span> è la proprietà di uno schema di cifratura i cui testi cifrati per un attaccante sono essenzialmente <span style="color:gold">indistinguibili</span> da stringhe di bit generate casualmente.
* Più precisamente, la proprietà viene descritta in termini di un <span style="color:gold">gioco a due</span> (Alice e Eva).
    * Eva produce <span style="color:gold">due messaggi in chiaro</span> e li sottopone ad Alice;
    * Alice decide, a caso e con uguale probabilità, <span style="color:gold">quale dei due cifrare</span>, e lo presenta ad Eva;
    * Alice vince il gioco se Eva non riesce ad indovinare quale messaggio sia stato cifrato con probabilità significativamente maggiore di $\frac{1}{2}$.
    * Nel concreto questo vuol dire che Eva, dal suo punto di vista, non ha strategie più efficaci rispetto a <span style="color:gold">tirare a caso</span>.
* Un secondo obiettivo è la <span style="color:gold">non malleabilità</span>.
* Uno schema di cifratura si definisce <span style="color:gold">non malleabile</span> se, dato un ciphertext $C_1$ corrispondente ad un plaintext $P_1$, risulta impossibile per Eva creare un secondo ciphertext $C_2$ corrispondente ad un plaintext $P_2$ che abbia una <span style="color:gold">qualche relazione</span> con $P_1$.
* Si può intuire come la non malleabilità sia una proprietà che ha a che vedere con l'<span style="color:gold">integrità</span> dei messaggi.
* A riguardo del protocollo one-time pad, possiamo affermare che esso, pur essendo sicuro dal punto di vista della indistiguibilità, risulta però un cifrario <span style="color:gold">malleabile</span> quando la chiave venga riutilizzata.  

### <span style="color:cyan">Cifrari a blocchi</span>
* I moderni cifrari simmetrici vengono detti <span style="color:gold">a blocchi</span> perché lavorano su blocchi di dati di lunghezza fissata.
* Indicheremo genericamente con <span style="color:gold">$B$</span> il numero di bit di un blocco, un valore che tipicamente si colloca nell'<span style="color:gold">intervallo da 64 a 256</span>.
* Se il plaintext ha lunghezza minore di $B$, ad esso vengono preliminarmente aggiunti <span style="color:gold">bit di padding</span> (riempimento).
* Viceversa, se il plaintext ha lunghezza superiore, esso viene suddiviso in più blocchi che vengono poi <span style="color:gold">cifrati separatamente</span>.
* Come già detto a suo tempo, il particolare algoritmo con cui i blocchi di uno stesso messaggio vengono cifrati, di modo che si possa poi ricostruire (in sede di decifrazione) il plaintext originale, viene detto <span style="color:gold">mode of operation</span>.
* Entrambi gli aspetti, padding e mode of operation, sono <span style="color:gold">tutt'altro che banali</span> ed errori di valutazione possono dare vantaggi decisivi ad un attaccante.

#### <span style="color:cyan">Scelta di $B$</span>
* I valori di $B$ indicati sopra derivano dalla necessità di bilanciare due esigenze contrapposte.
* Da un lato $B$ deve essere un valore piccolo allo scopo di incrementare l'efficienza facilitando implementazioni in <span style="color:gold">hardware dedicato</span>.
* D'altro canto, $B$ non può essere troppo piccolo per non incorrere in seri <span style="color:gold">rischi di sicurezza</span> e (ancora dal punto di vista dell'efficienza) ridurre l'overhead legato al trattamento di molti <span style="color:gold">messaggi corti</span>;
* In particolare, se $B$ fosse troppo piccolo, un attaccante potrebbe precompilare una tabella di $2^{B}$ posizioni in cui alla generica posizione $i$ è memorizzato il <span style="color:gold">plaintext che ha $i$ come ciphertext</span>.
* In tal caso, l'attacco (denominato <span style="color:gold">codebook attack</span>) corrisponderebbe poi ad un semplice <span style="color:gold">table lookup</span>.
* Se, ad esempio, $B$ fosse 32, la tabella avrebbe $2^{32}$ posizioni, ciascuna di 32 bit (4 byte), per un totale di $4\cdot2^{32}=16$GB, cioè sarebbe <span style="color:gold">del tutto gestibile</span>.
* Con $B=64$ le dimensioni salirebbero a $2^{67}\approx 1.48\cdot 10^{20}$ byte, dunque senza alcuna possibilità di precomputazione e memorizzazione.

#### <span style="color:cyan">Sull'attacco codebook</span>
* &Egrave; lecito chiedersi: Come può l'attaccante precomputare i ciphertext, visto che non dispone della chiave?
* Dopo aver ragionato sugli <span style="color:gold">obiettivi generali di sicurezza</span>, questa domanda ci porta a riflettere sui cosiddetti <span style="color:gold">modelli di attacco</span>.
* Un modello di attacco è essenzialente un'<span style="color:gold">ipotesi</span> su quelle che sono le armi in mano all'attaccante.
* Qui ragioneremo solo di modelli <span style="color:gold">black box</span>, in cui l'attaccante non ha accesso o non utilizza informazioni diverse da input e output (plaintext e ciphertext).
* Esistono anche tipi di attacco che sfruttano <span style="color:gold">fenomeni fisici</span> piuttosto che proprietà matematiche.
* Per quanto possa a prima vista apparire incredibile, informazioni come il tempo di esecuzione di operazioni su una chiave privata possono essere utilizzate per violare un critto-sistema.
* Attacchi di questo tipo, che però non potremo prendere in considerazione, vengono detti (attacchi) <span style="color:gold">side-channel</span>.

#### <span style="color:cyan">Attack model</span>
* L'assunzione più semplice è che l'attaccante <span style="color:gold">veda solo il ciphertext</span> e debba costruire l'attacco basandosi su quello.
* Questo modello di attaccante viene detto <span style="color:gold">Ciphertext-only attacker (COA)</span>
* COA è un modello in molti casi adeguato (almeno in crittografia simmetrica).
* Tuttavia, dal punto di vista del difensore, supporre che l'attaccante non conosca nulla al di fuori del testo cifrato può indurre ad <span style="color:gold">abbassare pericolosamente "la guardia"</span>.
* &Egrave; invece prudente assumere che l'attaccante sia <span style="color:gold">sempre un passo avanti</span>.
* In ordine di "potenza" crescente, il modello successivo è noto come <span style="color:gold">Known-plaintext attacker (KPA)</span>, in cui l'attaccante (preliminarmente all'attacco) è stato in grado di osservare un certo numero di ciphertext con la conoscenza dei <span style="color:gold">corrispondenti plaintext</span>.

* Nei modelli precedenti, l'attaccante è comunque <span style="color:gold">passivo</span> perché non è lui/lei che determina quali testi cifrati ha possibilità di analizzare.
* Modelli più potenti prevedono che l'attaccante possa <span style="color:gold">effettuare query</span>.
* Il primo di tali modelli, in ordine di potenza, è il <span style="color:gold">Chosen-plaintext attacker (CPA)</span>, in cui l'attaccante può eseguire le cosiddette <span style="color:gold">encryption query</span>, può cioè scegliere un testo in chiaro e ottenere il corripondente ciphertext.
* Si noti che questo (CPA) è il modello di riferimento nel caso della <span style="color:gold">crittografia asimmetrica</span>, in cui l'attaccante conosce la chiave di cifratura.
* L'ultimo modello prevede che l'attaccante possa eseguire, oltre alle encryption query, anche <span style="color:gold">decryption query</span>, possa cioè scegliere un ciphertext e vedere il corrispondente plaintext.
* Questo modello viene detto <span style="color:gold">Chosen-ciphertext attacker (CCA)</span>.

#### <span style="color:cyan">Uno schema astratto per descrivere i block cipher</span>

* Cifratura e decifrazione richiedono una chiave segreta $k$, la cui lunghezza tipicamente è nel range <span style="color:gold">fra 128 e 256 bit</span>.
* Sappiamo che, matematicamente, la cifratura è una <span style="color:gold">permutazione</span> poiché plaintext e ciphertext sono composti dallo stesso numero di bit.
* Ad ogni possibile plaintext di $B$ bit deve cioè corrispondere <span style="color:gold">un unico ciphertext di $B$ bit</span>, e viceversa, altrimenti il procedimento non sarebbe <span style="color:gold">reversibile</span>.
* Può essere istruttivo visualizzare la cifratura come un <span style="color:gold">processo in due fasi</span>:
    1. data la chiave $k$, si seleziona una specifica <span style="color:gold">lookup table $T_k$</span>, di $2^B$ posizioni, che memorizza una specifica permutazione delle sequenze di $B$ bit;
    2. la cifratura corrisponde ad un accesso a $T_k$, usando il <span style="color:gold">plaintext come chiave</span>, per recuperare il corrispondente ciphertext.
* Come semplice esempio, con $B=3$, a <span style="color:gold">due differenti chiavi</span> potrebbero corrispondere le seguenti tabelle.
<table style="margin: 0 auto; border: 1px solid black;border-collapse: collapse;">
    <tr>
        <td style="border: 1px solid black;">
            \begin{array}{|r|r|c|c|}
            \hline
        \mathrm{P\ base\ 10}&\mathrm{P\ base\ 2}&\mathrm{C}\\
        \hline
        0_{10}&000&011\\
        \hline
        1_{10}&001&010\\
        \hline
        2_{10}&010&111\\
        \hline
        3_{10}&011&110\\
        \hline
        4_{10}&100&001\\
        \hline
        5_{10}&101&000\\
        \hline
        6_{10}&110&101\\
        \hline
        7_{10}&111&100\\
        \hline
\end{array}
        </td>
        <td bstyle="border: 0px;">&nbsp;&nbsp;</td>
        <td style="border: 1px solid black;">
            \begin{array}{|r|r|c|c|c|}
        \hline
        \mathrm{P\ base\ 10}&\mathrm{P\ base\ 2}&\mathrm{C}\\
        \hline
        0_{10}&000&100\\
        \hline
        1_{10}&001&001\\
        \hline
        2_{10}&010&000\\
        \hline
        3_{10}&011&111\\
        \hline
        4_{10}&100&110\\
        \hline
        5_{10}&101&011\\
        \hline
        6_{10}&110&010\\
        \hline
        7_{10}&111&101\\
        \hline
\end{array}
        </td>
    </tr>
</table>
<p />
* Naturalmente l'unica informazione delle due tabelle che viene effettivamente memorizzata è la <span style="color:gold">colonna C</span> (ciphertext). Le prime due colonne sono gli <span style="color:gold">indici</span>, rispettivamente in base 10 e in base 2, e rappresentano entrambe il plaintext.

#### <span style="color:cyan">Una riflessione su chiavi e permutazioni</span>
* Utilizzando il modello astratto possiamo ragionare quantitativamente sulle <span style="color:gold">permutazioni effettivamente utilizzabili</span> in un cifrario a blocchi e iniziare una riflessione su quali concretamente impiegare (e quali evitare).
* Sulla base del principio (che abbiamo enunciato in precedenza) in base al quale a chiavi diverse devono corrispondere permutazioni diverse, è immediato osservare che la lunghezza della chiave determina il <span style="color:gold">numero di lookup table</span>, e dunque di permutazioni, che possono essere utilizzate in un cifrario a blocchi con chiavi di quella lunghezza.
* Se dunque la chiave è lunga $k$ bit, il <span style="color:gold">numero di permutazioni usabili è $2^k$</span>.
* Questo valore, apparentemente "grande" (perché esponenziale nella lunghezza della chiave), deve essere visto sullo sfondo del <span style="color:gold">numero totale di permutazioni di $B$ bit</span>.
* Quest'ultima è una quantità che cresce molto più velocemente con $B$: essa è infatti pari a $2^B!$ (cioè il <span style="color:gold">fattoriale di $2^B$</span>).
* &Egrave; interessante verificare quanto dovrebbe essere lunga la chiave per poter utilizzare <span style="color:gold">tutte le permutazioni</span>, in funzione di valori crescenti di $B$.
* Con l'aiuto di Python, analizziamo i <span style="color:gold">casi $2\leq B\leq 7$</span> (cioè per blocchi di lunghezza così corta da essere del tutto improponibili in un cifrario reale).

In [ ]:
from math import factorial,log,log10
print("Valore di B\tNumero di permutazioni\tLunghezza minima della chiave")
for B in range(2,8):
    v = factorial(2**B)
    if B>3:
        e = int(log10(v))
        m = int(10*(v*10**(-e)))/10
        s = f"{m}e{e}"
    else:
        s = str(v)
    print(f"{B}\t\t{s}\t\t\t{int(B*2**B*log(2))}")

Valore di B	Numero di permutazioni	Lunghezza minima della chiave
2		24			5
3		40320			16
4		2.0e13			44
5		2.6e35			110
6		1.2e89			266
7		3.8e215			621


* Interessante anche l'analisi in qualche modo <span style="color:gold">"inversa"</span>.
* Fissiamo cioè una dimensione realistica per la chiave e valutiamo la <span style="color:gold">frazione di permutazioni utilizzabili</span> (rispetto a tutte quelle possibili) al crescere della dimensione $B$ dei blocchi.
* Ad esempio, con <span style="color:gold">chiavi di 128 bit</span> possiamo usarle tutte solo se $B\leq 5$, dopodiché la frazione di permutazioni utilizzabili diviene rapidamente <span style="color:gold">trascurabile</span>.

In [ ]:
k = 128
for B in range(5,9):
    print(f"Frazione di permutazioni utilizzabili con chiavi di {k} bit e "+\
          f"blocchi di {B} bit: {min(1,2**k/factorial(2**B))}")

Frazione di permutazioni utilizzabili con chiavi di 128 bit e blocchi di 5 bit: 1
Frazione di permutazioni utilizzabili con chiavi di 128 bit e blocchi di 6 bit: 2.681776295311788e-51
Frazione di permutazioni utilizzabili con chiavi di 128 bit e blocchi di 7 bit: 8.824281449889046e-178
Frazione di permutazioni utilizzabili con chiavi di 128 bit e blocchi di 8 bit: 0.0


#### <span style="color:cyan">Quali permutazioni usare?</span>
* Abbiamo visto che delle permutazioni possibili se ne può usare solo una <span style="color:gold">minuscola frazione</span>.
* D'altra parte molte permutazioni <span style="color:gold">non andrebbero bene</span> comunque.
* Ad esempio, le 26 possibili permutazioni usate dal cifrario di Cesare, associate alle 26 possibili chiavi, sono semplici <span style="color:gold">rotazioni</span>, dunque di facile inversione.
* Delle due permutazioni di sequenze di tre bit presentate nelle tabella riportate sopra (che qui vengono nuovamente riprodotte per comodità) quella di destra non va sicuramente bene perché (quantomeno) <span style="color:gold">rivela sempre il bit meno significativo del plaintext</span>.

<table style="margin: 0 auto; border: 1px solid black;border-collapse: collapse;">
    <tr>
        <td style="border: 1px solid black;">
            \begin{array}{|r|r|c|c|}
            \hline
        \mathrm{P\ base\ 10}&\mathrm{P\ base\ 2}&\mathrm{C}\\
        \hline
        0_{10}&000&011\\
        \hline
        1_{10}&001&010\\
        \hline
        2_{10}&010&111\\
        \hline
        3_{10}&011&110\\
        \hline
        4_{10}&100&001\\
        \hline
        5_{10}&101&000\\
        \hline
        6_{10}&110&101\\
        \hline
        7_{10}&111&100\\
        \hline
\end{array}
        </td>
        <td bstyle="border: 0px;">&nbsp;&nbsp;</td>
        <td style="border: 1px solid black;">
            \begin{array}{|r|r|c|c|c|}
        \hline
        \mathrm{P\ base\ 10}&\mathrm{P\ base\ 2}&\mathrm{C}\\
        \hline
        0_{10}&000&100\\
        \hline
        1_{10}&001&001\\
        \hline
        2_{10}&010&000\\
        \hline
        3_{10}&011&111\\
        \hline
        4_{10}&100&110\\
        \hline
        5_{10}&101&011\\
        \hline
        6_{10}&110&010\\
        \hline
        7_{10}&111&101\\
        \hline
\end{array}
        </td>
    </tr>
</table>

* In generale non vanno bene permutazioni che possono essere espresse mediante <span style="color:gold">trasformazioni lineari affini invertibili</span> sul campo $\mathbf{Z}_2$
$$
xP+b=y(x)
$$
in cui $P$ è una matrice non-singolare di ordine $B$ su $\mathbf{Z}_2$, mentre $x$ (il plaintext), $b$ e $y(x)$ (il ciphertext) sono vettori riga di $B$ elementi, sempre con elementi in $\mathbf{Z}_2$.
* Ad esempio, <span style="color:gold">la matrice di sinistra</span> nelle tabelle è rappresentabile come trasformazione lineare affine con

$$
P=\begin{array}{ccc}
0&1&0\\
1&0&0\\
0&0&1
\end{array}
$$
e
$$
b=\begin{array}{c}0&1&1\end{array}
$$

* Se la trasformazione è realizzata come permutazione lineare affine è infatti possibile, nel modello CPA, ricostruire $b$ e $P$ e dunque <span style="color:gold">invertire la trasformazione</span> stessa.
* Infatti $b$ semplicemente è l'immagine di $O$ (vettore di tutti zeri).
$$y(O)=O\cdot P+b=O+b=b$$
ed è quindi sufficiente, per determinarlo, effettuare la query che chiede il <span style="color:gold">ciphertext corrispondente al blocco di tutti 0</span>.
* Una volta noto $b$, siamo in grado di determinare esplicitamente $P$ con le sole ulteriori query in cui i blocchi da cifrare sono <span style="color:gold">$e_i=(0,\ldots,0,\underset{\mathrm{indice}\ i}{1},0,\ldots,0)$</span>.
* Scrivendo tutte le $B$ query in forma matriciale abbiamo:
$$
I\cdot P = y(I)
$$
ovvero $P=y(I)$
* In altri termini, la <span style="color:gold">risposta alle $B$ query ordinate</span>, è proprio la matrice di trasformazione $P$, che possiamo efficientemente invertire.

#### <span style="color:magenta">Combinazioni lineari affini in $\mathbf{Z}_2$</span>

* Sul campo $\mathbf{Z}_2$ il numero di matrici invertibili è dato dalla seguente formula, di derivazione relativamente semplice

$$
\prod_{i=0}^{n-1} (2^n- 2^i) = (2^n-1)(2^n-2)(2^n-2^2)...(2^n-2^{n-1})
$$

* Il seguente codice Python ne implementa il calcolo.

In [ ]:
from math import factorial

In [ ]:
def PermInvGF2(n):
    v = 1
    pow = 1<<n
    for j in range(n):
        v *= (pow-(1<<j))
    return v

In [ ]:
B = 3
PermInvGF2(B)*(2**B)

1344

In [ ]:
factorial(2**B)

40320

* Ad esempio, la seguente permutazione <span style="color:gold">non è esprimibile</span> come combinazione lineare affine invertibile

<table style="margin: 0 auto; border: 1px solid black;border-collapse: collapse;">
    <tr>
        <td style="border: 1px solid black;">
            \begin{array}{|r|d|c|}
        \hline
        0_{10}&000&010\\
        \hline
        1_{10}&001&101\\
        \hline
        2_{10}&010&011\\
        \hline
        3_{10}&011&001\\
        \hline
        4_{10}&100&000\\
        \hline
        5_{10}&101&110\\
        \hline
        6_{10}&110&100\\
        \hline
        7_{10}&111&111\\
        \hline
\end{array}
        </td>
    </tr>
</table>

#### <span style="color:cyan">Proprietà della permutazioni usate in pratica</span>
* Ovviamente l'algoritmo di cifratura <span style="color:gold">non segue i passaggi indicati</span>, che sono solo (come detto) una maniera per descriverne il comportamento osservabile.
* Tuttavia, questa descrizione ha il pregio di mettere in chiaro come, nella realizzazione concreta di un cifrario a blocchi, debba necessariamente esistere un qualche passaggio algoritmico per <span style="color:gold">forzare non linearità</span> nella permutazione realizzata.
* Altre proprietà che devono essere soddisfatte dalle permutazioni sono <span style="color:gold">diffusione</span> e <span style="color:gold">confusione</span>.
* La diffusione richiede che ogni minima modifica, localizzata in un punto del plaintext si rifletta su <span style="color:gold">tutto il testo cifrato</span>.
* La confusione è la caratteristica che nel ciphertext <span style="color:gold">distrugge ogni regolarità</span>, ogni eventuale pattern (ad esempio due caratteri consecutivi identici) presente nel plaintext.
* Come vedremo presentando il <span style="color:gold">cifrario di Feistel</span>, il modo per realizzare algoritmicamente permutazioni con tutte le caratteristiche suddette è realizzarlo come <span style="color:gold">rete di componenti più semplici</span>.

### <span style="color:cyan">Feistel networks</span>
* Più che un cifrario, si tratta di uno schema per realizzare cifrari a blocchi, fra i quali il più famoso è ll <span style="font-style: italic; color:gold">Digital Encryption Standard</span> (<span style="font-style: italic; color:gold">DES</span>), per molti anni il "riferimento" nella cifratura simmetrica.
* Una rete di Feistel è composta da un certo numero di <span style="color:gold">stadi con identica struttura</span>.
* Ad ogni stadio viene utilizzata una specifica <span style="color:gold">round key $k_i$</span>; tutte le round key sono diverse e ottenute dalla <span style="color:gold">chiave generale</span> utilizzata dal cifrario mediante un algoritmo noto come <span style="color:gold">key schedule</span>.
* Il generico stadio $i$ opera su <span style="color:gold">due semiblocchi $L_i$ ed $R_i$</span> di dimensione $B/2$. Inizialmente, la <span style="color:gold">concatenazione $L_0R_0$ dei semiblocchi</span> è proprio il blocco completo da cifrare.
* Uno dei due blocchi, per la precisione il <span style="color:gold">blocco di destra $R_i$</span>, passa inalterato allo stadio successivo come prossimo blocco di sinistra, $L_{i+1}$.
* Invece, il successivo blocco di destra, $R_{i+1}$, è ottenuto <span style="color:gold">applicando ad $R_i$ una trasformazione $F$</span> dipendente dalla round key e il cui valore è posto in xor bitwise con $L_i$.

#### <span style="color:cyan">Schema grafico e corrispondenti equazioni</span>
* Il seguente è un classico schema per descrivere una Feistel network

<center>
    <img src="feistel.jpg" width="500">
</center>
                                       
* ... e queste le corrispondenti <span style="color:gold">equazioni che definiscono i blocchi</span> (parametrizzate rispetto ad $F$):                   

$$
\begin{eqnarray*}
L_1&=&R_0\\
R_1&=&L_0\oplus F(R_0,k_1)\\
L_2&=&R_1\\
R_2&=&L_1\oplus F(R_1,k_2)\\
&\ldots&\\
L_r&=&R_{r-1}\\
R_r&=&L_{r-1}\oplus F(R_{r-1},k_r)
\end{eqnarray*}
$$

* Il primo problema da risolvere è verificare la <span style="color:gold">reversibilità del processo</span>.
* Questo è immediato, grazie alle <span style="color:gold">proprietà dell'or exclusivo</span>
$$
\begin{eqnarray*}
R_{r-1}&=&L_r\\
L_{r-1}&=&R_r\oplus F(L_r,k_r)\\
R_{r-2}&=&L_{r-1}\\
L_{r-2}&=&R_{r-1}\oplus F(L_{r-1},k_{r-1})\\
&\ldots&\\
L_0&=&R_{1}\\
R_0&=&L_1\oplus F(R_{1},k_1)
\end{eqnarray*}
$$
* Questa seconda computazione, che calcola la decifrazione, può essere eseguita dalla rete di figura <span style="color:gold">"rovesciando" tutte le frecce</span> e dunque la direzione input/output dei semiblocchi.
* Tuttavia, questo <span style="color:gold">non è necessario</span>! &Egrave; sufficiente infatti alimentare la stessa rete di figura con input $R_rL_r$ e, naturalmente, usando le round key in ordine inverso.
* Il numero di round utilizzati in un cifrario reale deve essere tale da consentire ad una variazione nell'input di propagarsi a tutto il ciphertext (in altri termini, il numero di round è <span style="color:gold">legato alla diffusione</span>).

#### <span style="color:cyan">La funzione F</span>
<center>
        <img src="desround.jpg" width="400">
</center>

* La funzione $F$, i cui dettagli variano a seconda della particolare implementazione del cifrario (la figura mostra <span style="color:gold">il caso del DES</span>), ha comunque come scopo principale:
    1. di inserire <span style="color:gold">la chiave nel processo</span>, "mischiandola" con i bit del messaggio;
    2. assicurare <span style="color:gold">dipendenza non lineare</span> fra plaintext e ciphertext, attraverso l'S-box.
* Nel caso del DES i passaggi (da destra a sinistra in figura) sono i seguenti:
    1. I 32 bit di $R_i$ vengono <span style="color:gold">espansi</span> a 48 attraverso la duplicazione di alcuni bit;
    2. il risultato (48 bit) vengono messo in <span style="color:gold">xor con la round key</span> che, nel DES, è appunto di 48 bit;
    3. il nuovo risultato intermedio deve essere riportato a 32 bit, in vista dell'xor finale con il semiblocco $L_i$; questo è uno dei compiti dell'S-box, che è in realtà costituito da <span style="color:gold">8 tabelle, ciascuna indicizzata da 6 bit e le cui posizioni contengono 4 bit di output</span> (si veda la figura sotto);
    4. infine, il risultato prodotto dall'S-box ($4\cdot 8=32$ bit) è <span style="color:gold">permutato e posto in xor col semiblocco $L_i$</span> che, come sappiamo, costituirà il blocco destro del round successivo.


<center>
        <img src="sbox.jpg" width="800">
</center>
Una tabella componente l'S-box. Per selezionare la riga vengono usati 2 bit mentre 4 servono per la colonna. Il risultato è un numero fra 0 e 15 interpretato come sequenza di 4 bit.

#### <span style="color:cyan">Key schedule</span>
* L'algoritmo con cui, a partire dalla chave iniziale, vengono ottenute le varie round key è detto <span style="color:gold">key schedule</span>.
* Il DES usa una chiave "generale" di 56 bit (del tutto inadeguata oggi) che viene dapprime suddivisa in <span style="color:gold">due metà di 28 bit ciascuna</span>.
* Ad ogni round, ciascuna metà viene prima <span style="color:gold">ruotata a sinistra di 1 o 2 posizioni</span> (a seconda del round), dopodiché viene formata la round key prelevando <span style="color:gold">24 bit da entrambe</span>.

### <span style="color:cyan">Mode of operation</span>
* Il mode of operation è l'algoritmo che produce la <span style="color:gold">cifratura complessiva</span> di un messaggio utilizzando un cifrario a blocchi per le singole porzioni del messaggio.
* Dapprima viene effettuato il riempimento (<span style="color:gold">padding</span>) del messaggio iniziale $M$ di modo che la sua lunghezza sia un multiplo della dimensione dei blocchi
* Indicheremo con
$$
P = P_1 || P_2 || \ldots || P_n
$$
la suddivisione in blocchi del plaintext già dopo il padding.
* Considereremo, esclusivamente a modo di esempio, due tecniche differenti per il mode of operation.

#### <span style="color:cyan">Il modo ECB (Electronic Codebook Mode)</span>

* &Egrave; l'algoritmo più semplice ma anche il <span style="color:gold">più vulnerabile</span>.
* Ogni singolo blocco viene infatti cifrato in modo <span style="color:gold">indipendente</span> dagli altri ma usando la <span style="color:gold">stessa chiave</span>.
$$
C_i = E(P_i,K),\qquad i=1,2,\ldots,n
$$
dove chiaramente $E$ indica la funzione realizzata dal particolare cifrario a blocchi utilizzato e $K$ è la chiave.
* Il problema principale con ECB è che a blocchi di plaintext identici corrispondono <span style="color:gold">blocchi cifrati identici</span>.
* Non si pensi che questa sia una situazione molto improbabile.
* Come per il cifrario di Vigenère, se due blocchi di testo identici si <span style="color:gold">allineano alla dimensione del blocco</span>, i testi cifrati sono identici.
* Particolarmente problematico è il caso della cifratura di <span style="color:gold">immagini o grafici</span> (pensiamo agli "sfondi").
* Il seguente è un esempio famoso:

<br />
<center>
    <img src="penguin.jpg" width="350">
    Il pinguino di Linux: immagine originale e immagine cifrata con modo ECB
</center>
<center>
Source: J.P. Aumasson, Serious Cryptography, No Starch Press (2018)
</center>

#### <span style="color:cyan">Il modo CBC (Cipher Block Chaining)</span>

* La cifratura avviene nel modo seguente:
$$
C_1 = E(P_1\oplus IV,K),\qquad C_{i} = E(P_{i}\oplus C_{i-1},K), i=2,3,\ldots,n
$$
dove $IV$ indica un blocco iniziale detto appunto <span style="color:gold">initialization vector</span>.
* La decifrazione è possibile grazie ancora alle <span style="color:gold">proprietà dell'or esclusivo</span>:
$$
P_1 = D(C_1,K) \oplus IV,\qquad P_i=D(C_i,K)\oplus C_{i-1}
$$
* Il vettore di inizializzazione può essere scelto in vario modo ma la soluzione più comune è di utilizzare un <span style="color:gold">nonce</span>.

### <span style="color:magenta">Uno schema di padding e l'attacco "Padding oracle"</span>
* Poiché la lunghezza del plaintext $P$ da cifrare può non essere (e in generale non è) un multiplo della dimensione $B$ del blocco, è necessario (come già accennato) procedere ad una <span style="color:gold">estensione</span> del messaggio che ne porti la lunghezza ad un tale multiplo.
* In gergo tecnico, tale estensione viene chiamata <span style="font-style: italic; color:gold">padding</span> (riempimento).
* &Egrave; ovviamente necessario che il padding sia poi riconosciuto dal destinatario, cioè non confuso con il messaggio vero e proprio.
* Semplici soluzioni, come inserire una sequenza di zeri in coda, sono chiaramente destinate al fallimento.
* &Egrave; necessario che mittente e destinatario seguano un vero e proprio <span style="color:gold">protocollo</span>.

#### <span style="color:magenta">Lo schema di padding ISO/IEC 9797-1</span>

* Qui e nel resto di questa sezione sul padding supponiamo che i blocchi siano composti da 16 byte (128 bit), come nel caso dell'<span style="font-style: italic; color:gold">American Encryption Standard</span> (AES).
* Lo standard ISO/IEC 9797-1 per il padding prevede di operare come segue.
* Come prima cosa si calcola la lunghezza $\ell$ del plaintext $P$ (lunghezza <span style="color:gold">espressa in bit</span>).
* Si prepara quindi il plaintext modificato $P'$ inserendo, nell'ordine:
    1. Un primo blocco in cui viene scritta la lunghezza $\ell$ di $P$ utilizzando tutti i 16 byte del blocco (quindi inserendo eventuali <span style="color:gold">zeri non significativi</span> all'inizio).
    2. Il plaintext originale $P$.
    3. Tanti bit quanto serve per completare l'ultimo blocco (eventualmente nessuno, nel caso in cui il messaggio abbia lunghezza che è multiplo di 128).
* Ad esempio, per un plaintext $P$ lungo <span style="color:gold">232 bit</span>, la codifica sarebbe come segue (si tenga presente che <span style="color:gold">$232_{10}=\mathrm{e8}_{16}=11101000_2$</span>):
<center>
    $\underbrace{00\ldots 0}_{120}11101000\ P\ \underbrace{00\ldots 0}_{24}$
</center>

#### <span style="color:magenta">Lo schema di padding oggetto dell'attacco</span>
* Un differente schema di padding, utilzzato per portare l'attacco di cui parleremo subito dopo, prevede invece di <span style="color:gold">non inserire</span> il blocco iniziale con la lunghezza ma di utilizzare <span style="color:gold">differenti sequenze di riempimento</span>, a seconda della lunghezza del plaintext originale $P$.
* Più precisamente, lo schema prevede quanto segue, dove la lunghezza $\ell$ di $P$ è espressa non in bit bensì in <span style="color:gold">byte</span>.
    * Se $\ell=16k$, per un qualche $k>0$, si appende un intero blocco in fondo a $P$, formato da 16 byte identici, il cui valore numerico è precisamente 16 (ovvero $10_{16}$)
    <center>
    <span style="color:gold">$\underbrace{10_{16}10_{16}\ldots 10_{16}}_{16}$</span>
    </center>
    * Se $\ell=16k+1$ (cioè, nel caso in cui l'ultimo blocco di $P$ occupi un solo byte dei 16 disponibili), in coda all'ultimo blocco si aggiungono 15 byte identici, il cui valore numerico è 15 (ovvero $0f_{16}$):
    <center>
    <span style="color:gold">$\underbrace{0f_{16}0f_{16}\ldots 0f_{16}}_{15}$</span>
    </center>
    * Se $\ell=16k+2$ (cioè, nel caso in cui l'ultimo blocco di $P$ occupi due byte dei 16 disponibili), in coda all'ultimo blocco si aggiungono 14 byte identici, il cui valore numerico è 14 (ovvero $0e_{16}$):
    <center>
    <span style="color:gold">$\underbrace{0e_{16}0e_{16}\ldots 0e_{16}}_{14}$</span>
    </center>
    * ....
    * Se $\ell=16k+15$ (cioè, nel caso in cui l'ultimo blocco di $P$ occupi tutti i byte disponibili eccetto uno), in coda all'ultimo blocco si aggiunge il byte di valore numerico 1 (ovvero <span style="color:gold">$01_{16}$</span>).
* Il vantaggio di questo schema è che utilizza un blocco in più <span style="color:gold">solo se la lunghezza del plaintext (in byte) è già un multiplo della lunghezza del blocco</span>. E chiaramente è necessario farlo in questo caso.

* In sede di decifrazione, eliminare il padding è molto semplice.
* Si guarda l'<span style="color:gold">ultimo byte dell'ultimo blocco decifrato</span> e se ne intepreta il valore numerico.
* Se tale valore è $n$, $n\in\{1,16\}$, il padding è formato precisamente dagli <span style="color:gold">ultimi $n$ byte</span>, che vengono eliminati.
* In particolare, se $n=16$, si elimina del tutto l'<span style="color:gold">ultimo blocco</span>.

#### <span style="color:magenta">Condizioni per l'attacco</span>
* L'attacco funziona con lo schema di padding appena illustrato e con qualsiasi cifrario blocchi che operi unitamente al <span style="color:gold">mode of operation CBC</span>.
* Il modello di attacco è "di fatto" debole; Eva non necessità altro che la conoscenza del cypertext. Eva deve poter sì effettuare query di decifrazione, ma l'attacco è verso un server che tipicamente risponde...
* &Egrave; però enecessario che il software che effettua la decifrazione sollevi un'eccezione (cioè, segnali un errore) nel caso di <span style="color:gold">padding malformato</span> (cosa che faceva, ad esempio, SSL/TLS).
* Un padding può essere malformato se l'ultimo byte ha valore numerico <span style="color:gold">0 oppure maggiore di 16</span>.
* Se invece ha valore $j\in \{1,16\}$, è malformato nel caso in cui almeno uno dei <span style="color:gold">precedenti $j-1$ byte</span> non ha valore $j$
* Per comodità, schematizziamo gli ultimi due round della decifrazione con CBC. Non usiamo gli indici ma indichiamo con $C'$ e $C$ rispettivamente il penultimo e l'ultimo blocco cifrato cui corrispondono i blocchi in chiaro $P'$ e $P$.  

<center>
    <img src="CBC.jpg" width="300">
</center>

* Notiamo che se modifichiamo il blocco $C'$, in molti casi il blocco $P$ che verrà messo in chiaro avrà un <span style="color:gold">padding malformato</span>.
* Immaginiamo di fissare ciascun bit di $C'$ a 0 oppure ad 1 con uguale probabilità e in modo indipendente dagli altri bit.
* Per le proprietà dell'or esclusivo (già analizzate nel caso del one-time padding) anche <span style="color:gold">ciascun bit di $P$ sarà 0 oppure 1 con uguale probabilità e in modo indipendente dagli altri</span>.
* Possiamo quindi calcolare la probabilità che il padding sia <span style="color:gold">ben formato</span>.
* Perché questo si verifichi, deve accadere che l'ultimo byte, considerato come numero, sia 1 oppure che gli ultimi due byte siano entrambi 2, o gli ultimi tre siano entrambi 3 e così via.
* Data l'indipendenza e l'equiprobabilità dei bit, l'evento che gli ultimi $k$ byte valgano tutti $k$ è $\frac{1}{256^k}$ e dunque, la probabilità complessiva di padding ben formato è

<center>
    <span style="color:gold">$\sum\limits_{k=1}^{16}\frac{1}{256^k}$</span>
</center>

In [ ]:
sum([1/(256**k) for k in range(16,0,-1)])

0.00392156862745098

* Da cui ovviamente risulta che il padding sarà malformato con probabilità:

In [ ]:
1-sum([1/(256**i) for i in range(16,0,-1)])

0.996078431372549

* Come  esercizio è anche istruttivo <span style="color:gold">stimare</span> la probabilità con un <span style="color:gold">algoritmo Monte Carlo</span>.
* Il programma potrebbe essere meglio comprensibile dopo aver letto il successivo sulla randomness. L'algoritmo è tuttavia molto semplice da poter essere facilmente descritto.
    1. Viene generato un certo numero di sequenze composte da <span style="color:gold">16 numeri casuali</span> compresi fra 0 e 255 (i numeri interi positivi rappresentabili in binario su 8 bit).
    2. Si verifica che la sequenza includa un <span style="color:gold">padding ben formato</span> (ultimi $k$ numeri uguali a $k$, per un qualche valore di $k$ compreso fra 1 e 16).
    3. In caso affermativo, si conta <span style="color:gold">un caso favorevole</span>.
    4. Al termine, la probabilità è approssimata dal rapporto fra casi favorevoli e sequenze generate.
    5. Questo valore sarà un'approssimazione tanto migliore quanto più è alto il numero di sequenze "testate".

In [ ]:
from random import randint

In [ ]:
n = 1000000 # numero di blocchi casuali generati
blen = 16   # lunghezza, in byte, di B
wfp = 0     # variabile che conta i well-formed padding
for _ in range(n):
    B = [randint(0,255) for _ in range(blen)]   # Blocco casuale
    lastbyte = B[-1]
    if lastbyte == 0 or lastbyte>blen:
        continue
    if all([B[i] == lastbyte for i in range(blen-2,blen-lastbyte-1,-1)]):
        if lastbyte != 1:
            print(B)
        wfp += 1
print(wfp/n)

[13, 246, 20, 99, 200, 139, 115, 90, 156, 188, 239, 170, 159, 122, 2, 2]
[179, 194, 122, 46, 91, 95, 181, 194, 120, 178, 167, 147, 209, 215, 2, 2]
[70, 184, 186, 250, 44, 0, 169, 133, 88, 23, 173, 36, 238, 160, 2, 2]
[145, 95, 124, 215, 10, 24, 210, 76, 159, 79, 20, 154, 138, 110, 2, 2]
[142, 186, 150, 143, 141, 76, 58, 237, 199, 184, 1, 186, 29, 230, 2, 2]
[198, 2, 144, 182, 107, 127, 29, 67, 112, 200, 103, 107, 187, 123, 2, 2]
0.003885


* Come si vede dall'esempio (con le stampe non commentate) il <span style="color:gold">"grosso" della probabilità</span> che il padding "casuale" sia ben formato è dato dall'accadimento che l'<span style="color:gold">ultimo byte sia $01_{16}$</span>.
* D'altra parte è facile calcolare esplicitamente la probabilità condizionata che l'ultimo byte sia $01_{16}$ dato che il padding è ben formato (wfp).

<center>
    <span style="color:gold">pr[lastbyte=$01_{16}$|wfp] = $\frac{\mathrm{pr}[\mathrm{lastbyte}=01_{16}\wedge\mathrm{wfp}]}{\mathrm{pr}[\mathrm{wfp}]}=\frac{\mathrm{pr}[\mathrm{lastbyte}=01_{16}]}{\mathrm{pr}[\mathrm{wfp}]}=\frac{\frac{1}{256}}{\sum\limits_{i=1}^{16}\frac{1}{256^i}}$</span>
</center>

In [ ]:
(1/256)/sum([1/(256**i) for i in range(16,0,-1)])

0.99609375

#### <span style="color:magenta">L'attacco</span>
<center>
    <img src="CBC.jpg" width="200">
</center>

* Eva sceglie a caso tutti i byte di $C'$ <span style="color:gold">escluso l'ultimo</span>.
* L'ultimo byte viene posto al valore $00_{16}$ e viene <span style="color:gold">incrementato di una unità</span> ad ogni query.
* Se, per un dato valore di $xy_{16}$, non viene generato errore, per quanto detto sopra Eva può essere ragionevolmente sicura che <span style="color:gold">l'ultimo byte di $P$ sia $01_{16}$</span>.
* Risulta cioè <span style="color:gold">$xy_{16}\oplus x[15]=01_{16}$</span> e dunque, date le proprietà dell'or esclusivo, Eva può recuperare il valore dell'ultimo byte di $X$ nel modo seguente: <span style="color:gold">$x[15]=xy_{16}\oplus 01_{16}$</span>
* A questo punto Eva ripete il trucco alla luce della conoscendza di $x[15]$.
* Fissa cioè il valore di $C'[15]$ in modo tale che $C'[15]\oplus x[15]=02_{16}$, e dunque (sempre per le proprietà dello xor)
<center>
    <span style="color:gold">$C'[15]=x[15]\oplus 02_{16}$</span>
</center>
sceglie poi i primi quattordici byte di $C'$ in modo random e inizia a far variare il byte $C'[14]$ a partire da 0 con incrementi unitari.
* Per lo stesso argomento di sopra, quando, per un certo valore $rs_{16}$ assegnato a $C'[14]$, non viene segnalato errore è altamente probabile che il penultimo byte del testo decifrato sia $02_{16}$, da cui Eva ricava
<center>
    <span style="color:gold">$X[14]=rs_{16}\oplus 02_{16}$</span>
</center>
* A questo punto si può immaginare il seguito. Conoscendo i due byte $X[14$ e $X[15]$, Eva fissa i byte $C'[14$ e $C'[15]$ in modo tale che gli ultimi due byte del messaggio decodificato siano <span style="color:gold">$03_{16}03_{16}$</span>, sceglie a caso i primi tredici byte di $C'$ e comincia ad eseguire richieste facendo variare il terz'ultimo byte ($C'[13]$) a partire da $00_{16}$.
* Procedendo in questo modo Eva riesce a conoscere <span style="color:gold">interamente $X$</span> e, usando il valore $C'$ "vero", è in grado di ottenere il plaintext $P$

* Il procedimento può essere tranquillamente <span style="color:gold">iterato</span> per conoscere tutti i blocchi del plaintext.
* Per il penultimo blocco (la cui cifratura dipende dai blocchi precedenti ma <span style="color:gold">non dall'ultimo</span>) Eva non deve fare altro che considerare i blocchi cifrati escluso l'ultimo.
* Deve cioè fare <span style="color:gold">"come se"</span> il messaggio cifrato fosse composto solo dai primi $n-1$ blocchi e sottoporre questi (con le modifiche indicate in precedenza) al software ricevente (es. un server TLS).
* Per gli altri il discorso è analogo.